[14.0][ADD]sale_purchase_requisition

anddago78 opened the pull request:
This module create a relation with a quotation and a purchase agreement, a smart button is added in the quotation and in the purchase agreement, automatically, all the quotation lines are passed to the purchase agreement.

All details of this module are described in https://github.com/OCA/sale-workflow/issues/2414

MT-2554 @moduon @rafaelbn @Shide @yajo please review :)

Commit:
[14.0][ADD]sale_purchase_requisition

Shide commented on sale_purchase_requisition/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+    "depends": [
+        "sale_management",
+        "purchase_requisition",
+        "purchase",
+        "purchase_discount",
```
`purchase` dependency not needed because `purchase_requisition` has the dependency
`purchase_discount`, why this dependency?

Shide commented on sale_purchase_requisition/models/purchase_requisition.py:
```diff
@@ -0,0 +1,20 @@
+# Copyright 2022-2023 Moduon Team S.L. <info@moduon.team>
```
Only 2023
```suggestion
# Copyright 2023 Moduon Team S.L. <info@moduon.team>
```

Shide commented on sale_purchase_requisition/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+# Copyright 2023-2023 Moduon - Andrea Cattalani
```
Only 2023
```suggestion
# Copyright 2023 Moduon - Andrea Cattalani
```

Shide commented on sale_purchase_requisition/models/sale_order.py:
```diff
@@ -0,0 +1,74 @@
+    purchase_requisition_ids = fields.One2many(
+        comodel_name="purchase.requisition",
+        inverse_name="sale_order_id",
+        string="Acuerdo de compra",
```
```suggestion
        string="Purchase Requisitions",
```
on English, please

Shide commented on sale_purchase_requisition/models/sale_order.py:
```diff
@@ -0,0 +1,74 @@
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    purchase_agent = fields.Many2one(
+        comodel_name="res.users", string="Comercial compra"
+    )
```
It's already an user_id on the sale, why to add a new one?

If it's necessary, change the name to be in English

yajo replied:
This field is needed because the purchase agent might be different from the salesman.

However indeed it must change the name. Also, the field must follow the convention on naming:

```suggestion
    purchase_agent_id = fields.Many2one(
        comodel_name="res.users", string="Purchase agent"
    )
```

Shide commented on sale_purchase_requisition/models/sale_order.py:
```diff
@@ -0,0 +1,74 @@
+            "res_id": self.purchase_requisition_ids.id,
+        }
+
+    def _create_lines_preq(self):
```
Better rename to `_prepare_purchase_requisition_lines`.
Odoo standard of naming functions which returns prepared data for creations are called starting with `_prepare`.

Shide commented on sale_purchase_requisition/models/sale_order.py:
```diff
@@ -0,0 +1,74 @@
+        for rec in self:
+            if "order_line" in vals and rec.state in {"draft", "sent"}:
+                for pr in rec.purchase_requisition_ids:
+                    pr.line_ids.unlink()
```
This will block the order modifications because unlinking purchase requisition lines when are not in draft, throws an UserError: https://github.com/odoo/odoo/blob/14.0/addons/purchase_requisition/models/purchase_requisition.py#L157

Shide commented on sale_purchase_requisition/views/sale_order.xml:
```diff
@@ -0,0 +1,36 @@
+        <field name="name">sale.order.form.inherited.</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form" />
+        <field name="priority">10</field>
```
Why priority 10? Just remove it, don't needed

Shide commented on sale_purchase_requisition/views/purchase_requisition.xml:
```diff
@@ -0,0 +1,35 @@
+            name="inherit_id"
+            ref="purchase_requisition.view_purchase_requisition_form"
+        />
+        <field name="priority">10</field>
```
Remove low priority

Shide commented on sale_purchase_requisition/views/purchase_requisition.xml:
```diff
@@ -0,0 +1,35 @@
+            <xpath expr="//div[hasclass('oe_button_box')]" position="inside">
+                <button
+                    name="action_view_sale_order_requisition"
+                    string="Presupuesto"
```
On English
```suggestion
                    string="Quotation"
```

Shide commented on sale_purchase_requisition/views/purchase_requisition.xml:
```diff
@@ -0,0 +1,35 @@
+                <field
+                    name='sale_order_id'
+                    readonly="1"
+                    groups="sales_team.group_sale_salesman,sales_team.group_sale_manager"
```
Sale manager Group has salesman, so no need to duplicate.
```suggestion
                    groups="sales_team.group_sale_salesman"
```

Shide commented on sale_purchase_requisition/views/sale_order.xml:
```diff
@@ -0,0 +1,36 @@
+                    type="object"
+                    class="oe_stat_button"
+                    icon="fa-list-alt"
+                    string="Acuerdo de compra"
```
English
```suggestion
                    string="Purchase Requisition"
```

Shide commented on sale_purchase_requisition/views/sale_order.xml:
```diff
@@ -0,0 +1,36 @@
+            <xpath expr="//button[@name='action_confirm']" position="after">
+                <button
+                    name="action_create_purchase_requisition"
+                    string="Nuevo acuerdo de compra"
```
English
```suggestion
                    string="New Purchas Requisition"
```

Shide commented on sale_purchase_requisition/views/sale_order.xml:
```diff
@@ -0,0 +1,36 @@
+                    attrs="{'invisible': [('purchase_requisition_ids', '=', [])]}"
+                />
+            </xpath>
+            <xpath expr="//button[@name='action_confirm']" position="after">
```
Better use a generic locator
```suggestion
            <xpath expr="//form/header" position="inside">
```

Shide commented on sale_purchase_requisition/views/sale_order.xml:
```diff
@@ -0,0 +1,36 @@
+                    name="action_create_purchase_requisition"
+                    string="Nuevo acuerdo de compra"
+                    type="object"
+                    attrs="{'invisible': [('purchase_requisition_ids', '!=', [])]}"
```
Add state domain to "New Purchase Requisition" button
```suggestion
                    attrs="{'invisible': ['|', ('state', 'not in', ['draft', 'send']), ('purchase_requisition_ids', '!=', [])]}"
```

Shide commented on sale_purchase_requisition/models/sale_order.py:
```diff
@@ -0,0 +1,74 @@
+                        {
+                            "product_id": line.product_id.id,
+                            "product_qty": line.product_qty,
+                            "product_uom_id": line.product_uom.id,
```
Price unit is not propagated. Is it intended?

```suggestion
                            "product_uom_id": line.product_uom.id,
                            "price_unit": line.price_unit,
```

Shide reviewed (changes requested):
Good job! We can keep improving this a little bit.

Also I got an Access Error when creating a Sale Order with an user with only permissions on Sale: "Own documents only"

![image](https://user-images.githubusercontent.com/1162050/227483068-4fb4c3c5-070b-415e-ad48-65db20ebbfb8.png)

You will need to add a new Rule to let a salesman to read all of the Purchase Requisitions related to a Sale and Edit only the ones who he is the commercial/creator.

rafaelbn reviewed (commented):
Internal reviews. Changes needed 😄 @anddago78 

Missing

- [x] Sales man with Sale: "Own documents only" cannot create sale order as @Shide commented
- [x] Missing sales man NEW FIELDS in purchase requisition and propagate it from sales
- [x] Purchase Representative should be empty in creation to assign
- [ ] [FIX] Don't allow to change quantity in quotation once purchase requisition is confirmed (detailed description in MT-2554)
- [x] I'm a contributor 😃 from the very begging + Contributor are not in Moduon Standard 
- [x] Missing translations

yajo commented:
> [ ] [FIX ]Allowed to change quantity in quotation once purchase requisition is confirmed- Should allow

I don't understand this sentence.

yajo commented:
All comments attended, except the one I don't understand.

rafaelbn commented:
/ocabot rebase

rafaelbn commented:
@anddago78 @yajo test fails and runbot doesnt' build, what could we do? Thank you!

yajo commented on sale_purchase_requisition/models/sale_order.py:
You can see from codecov inline comments that there are a few lines in this file that aren't covered by tests.

Please add more test coverage.

yajo commented on sale_purchase_requisition/readme/USAGE.rst:
```diff
@@ -0,0 +1,5 @@
+To use this module, you need to:
+
+#. Install the module
+#. Automatically after save the new quotation apears a new button that we permit to create a new purchase agreements passing automatically all lines of your quotation.
```
```suggestion
#. Automatically after saving the new quotation, there will apear a new button that allows to create a new purchase agreement passing automatically all lines of your quotation.
```

yajo commented on sale_purchase_requisition/readme/USAGE.rst:
```diff
@@ -0,0 +1,5 @@
+
+#. Install the module
+#. Automatically after save the new quotation apears a new button that we permit to create a new purchase agreements passing automatically all lines of your quotation.
+#. If you add a value in the purchase_agent field this value will pass in the purchase agreement what you are creating.
```
```suggestion
#. If you add a value in the purchase agent field, this value will pass to the purchase agreement what you are creating.
```

yajo commented on sale_purchase_requisition/readme/ROADMAP.rst:
```diff
@@ -0,0 +1 @@
+When you modify the quotation lines the module cancel all purchase requisition lines and recreate them.
```
```suggestion
When you modify the quotation lines the module should cancel all purchase requisition lines and recreate them.
```

yajo commented on sale_purchase_requisition/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,3 @@
+This module connects a quotation with a purchase agreement.
+A smart button is added in the quotation and in the purchase agreement.
```
```suggestion
Smart buttons are added in the quotation and in the purchase agreement.
```

yajo commented on sale_purchase_requisition/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,3 @@
+This module connects a quotation with a purchase agreement.
+A smart button is added in the quotation and in the purchase agreement.
+Automatically, all the quotation lines are passed to the purchase agreement and the person in charge will be the purchase salesperson
```
```suggestion
Automatically, all the quotation lines are passed to the purchase agreement and the person in charge will be the purchase agent.
```

yajo commented on sale_purchase_requisition/models/sale_order.py:
```diff
@@ -0,0 +1,174 @@
+        inverse_name="sale_id",
+        string="Purchase requisition",
+    )
+    opened_purchase_requisition_count = fields.Integer(
+        compute="_compute_opened_purchase_requisition_count",
+        help="Number of purchase requisitions in open state",
+        string="Opened Purchase Requisitions",
+    )
```
[English is weird](https://dictionary.cambridge.org/pt/gramatica/gramatica-britanica/open-or-opened):

```suggestion
    open_purchase_requisition_count = fields.Integer(
        compute="_compute_opened_purchase_requisition_count",
        help="Number of purchase requisitions in open state",
        string="Open Purchase Requisitions",
    )
```

yajo commented on sale_purchase_requisition/models/sale_order.py:
```diff
@@ -0,0 +1,174 @@
+    purchase_requisition_ids = fields.One2many(
+        comodel_name="purchase.requisition",
+        inverse_name="sale_id",
+        string="Purchase requisition",
```
```suggestion
        string="Purchase requisitions",
```

yajo commented on sale_purchase_requisition/models/sale_order.py:
```diff
@@ -0,0 +1,174 @@
+                ).mapped("requisition_id")
+            if not_synced_prs:
+                sale.purchase_requisition_sync_warning = _(
+                    "Some lines are not synced on these Purchase Requisitions: %s",
```
I think it is particularly interesting that at least one test should assert the warning is being displayed appropriately.

yajo reviewed (changes requested):
The module looks very good. It just needs to polish some rough edges. Thanks!

yajo commented on sale_purchase_requisition/views/sale_order.xml:
```diff
@@ -0,0 +1,48 @@
+                    name="action_create_purchase_requisition"
+                    string="New Purchase Requisition"
+                    type="object"
+                    attrs="{'invisible': ['|', ('state', 'not in', ['draft', 'sent']), ('opened_purchase_requisition_count', '>', 0)]}"
```
I think this is making CI fail:

```suggestion
                    attrs="{'invisible': ['|', ('state', 'not in', ['draft', 'sent']), ('open_purchase_requisition_count', '>', 0)]}"
```

fcvalgar reviewed (approved):
Funtional review
LGTM thank you @anddago78

yajo commented on sale_purchase_requisition/models/purchase_requisition.py:
```diff
@@ -0,0 +1,54 @@
+                rec.purchase_requisition_sync_warning = _(
+                    "Some lines are not synced on these Sale order:"
+                    + "%s please update this lines manually" % (rec.sale_id.name),
+                )
```
This code contains a bug. You should never format the string before passing it to the translation function. It must be done after, always.

Besides, there are a couple of spell errors.

```suggestion
                rec.purchase_requisition_sync_warning = _(
                    "Some lines are not synced on this Sale order: %s. "
                    "Please update these lines manually.",
                    rec.sale_id.name,
                )
```

Shide commented on sale_purchase_requisition/security/ir.model.access.csv:
```diff
@@ -0,0 +1,7 @@
+access_purchase_requisition_type_sale_salesman,access_purchase_requisition_type_sale_salesman,purchase_requisition.model_purchase_requisition_type,sales_team.group_sale_salesman,1,0,0,0
+access_purchase_order_salesman,access_purchase_order_salesman,purchase.model_purchase_order,sales_team.group_sale_salesman,1,0,0,0
+access_mail_activity_salesman,mail.activity.salesman,mail.model_mail_activity,sales_team.group_sale_salesman,1,1,1,0
+access_mail_activity_type_salesman,mail.activity.type.salesman,mail.model_mail_activity_type,sales_team.group_sale_salesman,1,1,1,0
```
Why these permissions? Should be removed because everyone can see and edit activities if at least are assigned to it.

Shide reviewed (changes requested):
Permissions not needed

yajo commented on sale_purchase_requisition/models/purchase_requisition.py:
```diff
@@ -0,0 +1,76 @@
+        string="Sale Representative",
+        related="sale_id.user_id",
+    )
+    alert_activity = fields.Boolean(default=False)
```
Remove this field (see below).

yajo commented on sale_purchase_requisition/models/purchase_requisition.py:
```diff
@@ -0,0 +1,76 @@
+                    "Some lines are not synced on these Sale order:"
+                    + "%s please update this lines manually" % (rec.sale_id.name),
+                )
+                if (
+                    not rec.alert_activity
+                    and len(self) == 1
+                    and (rec.user_id or rec.sale_user_id)
+                ):
```
It makes no sense to check `len(self) == 1` because it's always gonna be. In this method's 1st line you're iterating the records already.

yajo commented on sale_purchase_requisition/models/purchase_requisition.py:
```diff
@@ -0,0 +1,76 @@
+                    and len(self) == 1
+                    and (rec.user_id or rec.sale_user_id)
+                ):
+                    self.env["mail.activity"].sudo().create(
+                        {
+                            "activity_type_id": self.env.ref(
+                                "mail.mail_activity_data_todo"
+                            ).id,
+                            "user_id": rec.user_id.id or rec.sale_user_id.id,
+                            "res_id": rec.id,
+                            "res_model_id": self.env["ir.model"]
+                            .search([("model", "=", "purchase.requisition")], limit=1)
+                            .id,
+                            "summary": _("Resolve differences with the sales order"),
+                        }
+                    )
```
OTOH, it's theoretically incorrect to create a mail activity in a compute mehtod. It works, but it shouldn't be done like this. Compute methods exist only to compute the value of a field, not to create other records.

Thus, this logic should be handled in the `sale.order` model, overriding `write()` and `create()`.

You can also simplify the creation of the activity by using helper methods. 

An example:

```python
class SaleOrder(models.Model):
    ...
    def write(self, values):
        result = super().write(values)
        for requisition in self.filtered("purchase_requisition_sync_warning"):
            requisition.activity_reschedule(
                ["sale_purchase_requisition.activity_difference"], 
                date_deadline=datetime.now(), 
                new_user_id=(requisition.user_id or requisition.sale_user_id).id
            ) or requisition.activity_schedule(
                "sale_purchase_requisition.activity_difference",
                date_deadline=datetime.now(), 
                user_id=(requisition.user_id or requisition.sale_user_id).id,
                summary=_("Resolve differences with the sales order"),
            )
        return result
```

yajo commented on sale_purchase_requisition/models/purchase_requisition.py:
```diff
@@ -0,0 +1,76 @@
+            if "product_qty" in values and values["product_qty"] != rec.product_qty:
+                rec.requisition_id.message_post(
+                    body=_(
+                        "The product quantity %s <br> have changed %d -> %d %s"
```
When a translation string contains more than 1 placeholder, you must use named placeholders:

```suggestion
                        "The product quantities %(product)s <br/> have changed %(old_qty)d -> %(new_qty)d %(uom)s"
```

This is because different languages might need to position words in different orders.

yajo commented on sale_purchase_requisition/models/purchase_requisition.py:
```diff
@@ -0,0 +1,76 @@
+                rec.requisition_id.message_post(
+                    body=_(
+                        "The product quantity %s <br> have changed %d -> %d %s"
+                        % (
```
As explained above, you should never format the string *before* translation. Always after.

yajo commented on sale_purchase_requisition/models/purchase_requisition.py:
```diff
@@ -0,0 +1,76 @@
+                    body=_(
+                        "The product quantity %s <br> have changed %d -> %d %s"
+                        % (
+                            rec.product_id.name,
+                            rec.product_qty,
+                            values["product_qty"],
+                            rec.product_uom_id.name,
```
To be compatible with named placeholders, now this must be a kwarg.

yajo reviewed (changes requested):
Some changes still needed, sorry. Thanks for your patience! 😊

anddago78 commented:
Thanks @yajo all comments are attended

Shide commented on sale_purchase_requisition/models/sale_order.py:
```diff
@@ -0,0 +1,218 @@
+        purchase_requisitions.filtered(lambda pr: pr.state == "open").action_done()
+        return res
+
+    def _unsynced_prs(self):
+        """Get purchase requisitions that are out-of-sync with sale orders."""
+        result = self.env["purchase.requisition"].browse()
+        for line in self.order_line.filtered("product_id.purchase_ok"):
+            result |= line.purchase_requisition_line_ids.filtered(
+                # Convert quantity to the same UoM as the line
+                lambda prl: prl.requisition_id.state != "cancel"
+                and prl.product_uom_id._compute_quantity(
+                    prl.product_qty, line.product_uom
+                )
+                != line.product_uom_qty
+            ).mapped("requisition_id")
+            for prid in self.purchase_requisition_ids:
+                compare_lines_so = (
+                    self.order_line.filtered("product_id.purchase_ok")
+                    .mapped("product_id")
+                    .ids
+                )
+                compare_lines_po = (
+                    self.env["purchase.requisition"]
+                    .browse(prid.id)
+                    .line_ids.mapped("product_id")
+                    .ids
+                )
+                set_lines_so = set(compare_lines_so).symmetric_difference(
+                    set(compare_lines_po)
+                )
+                if set_lines_so:
+                    result |= prid
+        return result
```
Avoid using nested for, better naming and check every sale separated.
```suggestion
    @api.returns('purchase.requisition', lambda value: value.id)
    def _unsynced_prs(self):
        """Get purchase requisitions that are out-of-sync with sale orders."""
        unsynced_prs = self.env["purchase.requisition"].browse()

        for sale in self:
            # Check if all products are in sale and purchase requisition
            so_p_ids = set(sale.order_line.mapped("product_id").filtered("purchase_ok").ids)
            for purchase_requisition in sale.purchase_requisition_ids:
                if so_p_ids ^ set(purchase_requisition.line_ids.mapped("product_id").ids):
                    # Symetric difference
                    unsynced_prs |= purchase_requisition

            # Check if quantities are the same in sale line and purchase requisition line
            lines_to_check = sale.order_line.filtered(
                # Exclude lines that cannot be purchase or already has unsynced prs
                lambda sol: sol.product_id.purchase_ok and sol.purchase_requisition_line_ids.requisition_id not in unsynced_prs
            )
            for line in lines_to_check:
                unsynced_prs |= line.purchase_requisition_line_ids.filtered(
                    # Convert quantity to the same UoM as the line
                    lambda prl: prl.requisition_id.state != "cancel"
                        and prl.product_uom_id._compute_quantity(
                            prl.product_qty, line.product_uom
                        )
                        != line.product_uom_qty
                ).mapped("requisition_id")
        return unsynced_prs
```

Shide reviewed (approved):
Just a small refactor to avoid nested `for` and fast execution

anddago78 commented:
thanks @Shide I have merged your change commented..

rafaelbn reviewed (changes requested):
> [FIX] Don't allow to change quantity in quotation once purchase requisition is confirmed (detailed description in [Moduon - Task #2554](https://www.moduon.team/web#id=2554&model=project.task&view_type=form))

The purchase user, still don't notice at all in any way that the sale order and purchase requisition are un synced, salemas has an red alert in sale but purchase user nothing

Detailed explanation in private with a long video

yajo committed:
[REF] sale_purchase_requisition: usability improvements

- Don't depend on `sale_management`, as that module adds no required features. Instead, just recommend installing it to be able to use those features.
- Rename purchase requisition sync warning field and compute method to make it more meaningful and standard.
- Only let salesperson modify purchase requisition while it's in draft.
- Move tests to `SavepointCase`, for performance.
- Let those salesperson that can see all leads interact with all purchase requisitions related to a sales order, just like if they were the assigned salesperson.
- Increase test coverage with a full flow to check sync and unsync behaviors.

@moduon MT-2554

yajo commented:
Pushed some improvements:

- Don't depend on `sale_management`, as that module adds no required features. Instead, just recommend installing it to be able to use those features.
- Rename purchase requisition sync warning field and compute method to make it more meaningful and standard.
- Only let salesperson modify purchase requisition while it's in draft.
- Move tests to `SavepointCase`, for performance.
- Let those salesperson that can see all leads interact with all purchase requisitions related to a sales order, just like if they were the assigned salesperson.
- Increase test coverage with a full flow to check sync and unsync behaviors.

yajo committed:
[FIX] sale_purchase_requisition: do not restrict permissions on full managers

If some user had both sale and purchase permissions, since Odoo didn't provide any rules for purchase users using purchase models, only the sale rules were used and we were effectively giving less access to people with more access.

@moduon MT-2554

rafaelbn reviewed (approved):
/ocabot merge patch
